query/src/sql/planner/binder/aggregate.rs

```diff
@@ -12,114 +12,471 @@
+                        None,
+                        alias,
+                        scalar.data_type(),
+                        Some(Box::new(scalar.clone())),
+                    );
+                    bind_context
+                        .aggregate_info
+                        .group_items
+                        .push(column_binding.clone());
+                    group_items.push(
+                        BoundColumnRef {
+                            column: column_binding.clone(),
+                        }
+                        .into(),
+                    );
+                    entry.insert(bind_context.aggregate_info.group_items.len() - 1);
                 }
+                continue;
             }
+
+            // Resolve scalar item and alias item
+            let scalar_binder = ScalarBinder::new(bind_context, self.ctx.clone());
+            let (scalar_expr, data_type) = scalar_binder
+                .bind_expr(expr)
+                .await
+                .or_else(|e| Self::resolve_alias_item(expr, &available_aliases, e))?;
+
+            if bind_context
+                .aggregate_info
+                .group_items_map
+                .get(&format!("{:?}", &scalar_expr))
+                .is_some()
+            {
+                // The group key is duplicated
+                continue;
+            }
+
+            let group_item_name = format!("{:#}", expr);
+            let index = if let Scalar::BoundColumnRef(BoundColumnRef {
+                column: ColumnBinding { index, .. },
+            }) = &scalar_expr
+            {
+                *index
+            } else {
+                self.metadata
+                    .add_column(group_item_name.clone(), data_type.clone(), None)
+            };
+
+            let column_binding = ColumnBinding {
+                table_name: None,
+                column_name: group_item_name.clone(),
+                index,
+                data_type,
+                scalar: Some(Box::new(scalar_expr.clone())),
+            };
+            bind_context
+                .aggregate_info
+                .group_items
+                .push(column_binding.clone());
+            bind_context.aggregate_info.group_items_map.insert(
+                format!("{:?}", &scalar_expr),
+                bind_context.aggregate_info.group_items.len() - 1,
+            );
+
+            group_items.push(
+                BoundColumnRef {
+                    column: column_binding,
+                }
+                .into(),
+            );
         }
 
-        Ok(AggregateInfo {
-            aggregate_functions,
-            available_aliases,
-        })
+        // Generate output columns of aggregate
+        self.build_aggregate_output_columns(bind_context)?;
+
+        // Enter in_grouping state
+        bind_context.in_grouping = true;
+
+        // Build a ProjectPlan, which will produce aggregate arguments and group items
```

Comment:
So is this the logical equivalent of adding `ExpressionTransform` to `build_aggregate` originally?

Reply:
Yes, this is not necessary in fact, we can remove this after refactoring the executor.